### PR TITLE
retrieve_utils.py - Updated.py to have the ability to parse text from PDF Files

### PR DESCRIPTION
**Pull request by AaronWard**
## Summary
This PR introduces a change to `retrieve_utils.py` to add the functionality to parse text from `.pdf` files. This is crucial for instantiating the `Chromadb` with `RetrieveUserProxyAgent` when there are `.pdf` files present in the docs directory.

- OS: Mac OS Ventura 13.5
- Python:  3.10.12
- autogen: 0.1.3
- scikit-learn: 1.3.1

## Motivation
I was encountering issues as the docs directory had `.pdf` files, and the existing constants in `retrieve_utils.py` did not support `.pdf` file formats. This enhancement will allow the `create_vector_db_from_dir` to handle `.pdf` files and prevent them being ignored as pdf's cannot be opened with `f.read()`

## Related issue number

#41

## Additional Information
I have also added the the ability to allow store documents if there chunk size < 40000, the reason for this being you may have only a small number of documents you want to query. I was unable to store because the chunk size was too small. 
If there is a specific reason for the 40000, please let me know and i'll remove.

```python
 # Upsert in batch of 40000 or less if the total number of chunks is less than 40000
  for i in range(0, len(chunks), min(40000, len(chunks))):
      end_idx = i + min(40000, len(chunks) - i)
      collection.upsert(
          documents=chunks[i:end_idx],
          ids=[f"doc_{j}" for j in range(i, end_idx)],  # unique for each doc
      )
```

## Changes Made
- Updated `retrieve_utils.py` to enable parsing text from PDF files.
- Updated `retrieve_utils.py` to allow for smaller chunksize vectorization for when you have few documents. 
- Updated retrieve chat example notebook to clarify what filetypes are accepted in `docs_path`.
- Added unit tests for `retrieve_utils.py`
  

## Checks
- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by AaronWard** [comment]
@microsoft-github-policy-service agree

File: autogen/retrieve_utils.py [hunk 1]
```diff
@@ -200,26 +237,18 @@ def create_vector_db_from_dir(
             collection_name,
             get_or_create=get_or_create,
             embedding_function=embedding_function,
-            # https://github.com/nmslib/hnswlib#supported-distances
```
Comment (thinkall) [inline autogen/retrieve_utils.py:203]: Could we keep these info to hnswlib, maybe in the docstring if you prefer not have them in the comments?

File: setup.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
     "diskcache",
     "termcolor",
     "flaml",
+    "PyPDF2",
```
Comment (thinkall) [inline setup.py:21]: It would be better to put this in the extras_require of `retrievechat`

**Review by thinkall — CHANGES_REQUESTED** [review]
Thank you so much for the contribution. Very good catch of the bug in the batch upsert! Overall, it looks good to me, just left two comments.

**Comment by thinkall** [comment]
The code formatting is failed, please add pre-commit-check by run `pre-commit install` in the folder of your local repo.

And the code coverage is decreased, maybe adding a test for the pdf extraction will address this issue.

File: setup.py (hunk 2 above)
Comment (AaronWard) [reply setup.py:21]: Done 👍

File: autogen/retrieve_utils.py (hunk 1 above)
Comment (AaronWard) [reply autogen/retrieve_utils.py]: My bad, didn't mean to remove this. They have been re-added.

**Comment by AaronWard** [comment]
> Thank you so much for the contribution. Very good catch of the bug in the batch upsert! Overall, it looks good to me, just left two comments.

Happy to help. I've run the `pre-commit install` command for formatting. I've gone ahead and wrote unit tests for `retrieve_utils.py`. I've repleaced `PyPDF2` with `pypdf` as the former is deprecated.

 Let me know if this PR is good to merge.

File: notebook/agentchat_RetrieveChat.ipynb [hunk 3]
```diff
@@ -148,7 +148,30 @@
...
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Accepted file formats for `docs_path`:\n",
+      "['txt', 'json', 'csv', 'tsv', 'md', 'html', 'htm', 'rtf', 'rst', 'jsonl', 'log', 'xml', 'yaml', 'yml']\n"
+     ]
```
Comment (AaronWard) [inline notebook/agentchat_RetrieveChat.ipynb:160]: This cell will need to be rerun after the merge to main to include `pdf` (if you want to keep it up to date).

File: autogen/retrieve_utils.py [hunk 4]
```diff
@@ -119,15 +137,51 @@ def split_text_to_chunks(
...
 
+def extract_text_from_pdf(file: str) -> str:
+    """Extract text from PDF files"""
+    text = ""
+    with open(file, "rb") as f:
+        reader = pypdf.PdfReader(f)
+        if reader.is_encrypted:  # Check if the PDF is encrypted
+            try:
+                reader.decrypt("")
+            except Exception as e:
```
Comment (thinkall) [inline autogen/retrieve_utils.py:148]: Is it possible to use a specific exception?

File: notebook/agentchat_RetrieveChat.ipynb [hunk 5]
```diff
@@ -148,7 +148,30 @@
...
-   "execution_count": 2,
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Accepted file formats for `docs_path`:\n",
+      "['txt', 'json', 'csv', 'tsv', 'md', 'html', 'htm', 'rtf', 'rst', 'jsonl', 'log', 'xml', 'yaml', 'yml']\n"
```
Comment (thinkall) [inline notebook/agentchat_RetrieveChat.ipynb:159]: If you run this notebook with your branch of code, `pdf` will be printed.

**Review by thinkall — COMMENTED** [review]
Thank you @AaronWard, it's almost good to merge. I've just two small suggestions to add. Thank you very much again for your nice PR and your patience!

File: autogen/retrieve_utils.py (hunk 4 above)
Comment (AaronWard) [reply autogen/retrieve_utils.py:148]: Done (see here for details: https://github.com/py-pdf/pypdf/blob/30db3efce36430bdbfd3412f904693c03b03daa0/pypdf/errors.py#L46C1-L47C1)

**Review by thinkall — APPROVED** [review]

**Comment by sonichi** [comment]
@AaronWard thanks for your contribution! Are you on discord?

**Comment by AaronWard** [comment]
> @AaronWard thanks for your contribution! Are you on discord?

no problem at all. no i don't but i just made and account - ill join server from the link in the repo

**Pull request merged**